partiql-types/src/main/kotlin/org/partiql/value/impl/PartiQLValueIonWriter.kt

```diff
@@ -0,0 +1,250 @@
+package org.partiql.value.impl
+
+import com.amazon.ion.Decimal
+import com.amazon.ion.IonSystem
+import com.amazon.ion.IonWriter
+import com.amazon.ion.system.IonBinaryWriterBuilder
+import com.amazon.ion.system.IonTextWriterBuilder
+import com.amazon.ion.system.IonWriterBuilder
+import com.amazon.ionelement.api.IonElement
+import com.amazon.ionelement.api.field
+import com.amazon.ionelement.api.ionBlob
+import com.amazon.ionelement.api.ionBool
+import com.amazon.ionelement.api.ionClob
+import com.amazon.ionelement.api.ionDecimal
+import com.amazon.ionelement.api.ionFloat
+import com.amazon.ionelement.api.ionInt
+import com.amazon.ionelement.api.ionListOf
+import com.amazon.ionelement.api.ionNull
+import com.amazon.ionelement.api.ionSexpOf
+import com.amazon.ionelement.api.ionString
+import com.amazon.ionelement.api.ionStructOf
+import com.amazon.ionelement.api.ionSymbol
+import org.partiql.value.Annotations
+import org.partiql.value.AnyValue
+import org.partiql.value.BagValue
+import org.partiql.value.BinaryValue
+import org.partiql.value.BlobValue
+import org.partiql.value.BoolValue
+import org.partiql.value.ByteValue
+import org.partiql.value.CharValue
+import org.partiql.value.ClobValue
+import org.partiql.value.CollectionValue
+import org.partiql.value.DateValue
+import org.partiql.value.DecimalValue
+import org.partiql.value.Float32Value
+import org.partiql.value.Float64Value
+import org.partiql.value.Int16Value
+import org.partiql.value.Int32Value
+import org.partiql.value.Int64Value
+import org.partiql.value.Int8Value
+import org.partiql.value.IntValue
+import org.partiql.value.IntervalValue
+import org.partiql.value.ListValue
+import org.partiql.value.MissingValue
+import org.partiql.value.NullValue
+import org.partiql.value.NumericValue
+import org.partiql.value.PartiQLValue
+import org.partiql.value.PartiQLValueBaseVisitor
+import org.partiql.value.PartiQLValueWriter
+import org.partiql.value.ScalarValue
+import org.partiql.value.SexpValue
+import org.partiql.value.StringValue
+import org.partiql.value.StructValue
+import org.partiql.value.SymbolValue
+import org.partiql.value.TextValue
+import org.partiql.value.TimeValue
+import org.partiql.value.TimestampValue
+import java.io.OutputStream
+
+internal class PartiQLValueIonWriter(
+    private val ionWriter: IonWriter,
+) : PartiQLValueWriter {
```

Comment:
I don't think we need to ever failed on writing to Ion. `$bag::[]` is a valid generic ion value, it is during reading we need to concern.